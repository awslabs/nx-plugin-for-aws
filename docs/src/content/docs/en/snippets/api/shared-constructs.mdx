---
title: API Shared Constructs
---
import { FileTree } from '@astrojs/starlight/components';
import Infrastructure from '@components/infrastructure.astro';

For deploying your API, the following files are generated:

<Infrastructure>
<Fragment slot="cdk">
<FileTree>
  - packages/common/constructs/src
    - app
      - apis
        - \<project-name>.ts CDK construct for deploying your API
    - core
      - api
        - http-api.ts CDK construct for deploying an HTTP API (if you selected to deploy an HTTP API)
        - rest-api.ts CDK construct for deploying a REST API (if you selected to deploy a REST API)
        - utils.ts Utilities for the API constructs
</FileTree>
</Fragment>
<Fragment slot="terraform">
<FileTree>
  - packages/common/terraform/src
    - app
      - apis
        - \<project-name>
          - \<project-name>.tf Module for deploying your API
    - core
      - api
        - http-api
          - http-api.tf Module for deploying an HTTP API (if you selected to deploy an HTTP API)
        - rest-api
          - rest-api.tf Module for deploying a REST API (if you selected to deploy a REST API)
</FileTree>
</Fragment>
</Infrastructure>