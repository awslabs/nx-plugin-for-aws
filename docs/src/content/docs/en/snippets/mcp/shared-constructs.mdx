---
title: MCP Shared Constructs
---
import { FileTree } from '@astrojs/starlight/components';
import Infrastructure from '@components/infrastructure.astro';

For deploying your MCP Server, the following files are generated:

<Infrastructure>
<Fragment slot="cdk">
<FileTree>
  - packages/common/constructs/src
    - app
      - mcp-servers
        - \<project-name>
          - \<project-name>.ts CDK construct for deploying your MCP Server
          - Dockerfile Passthrough docker file used by the CDK construct
    - core
      - agent-core
        - runtime.ts Generic CDK construct for deploying to Bedrock AgentCore Runtime
</FileTree>
</Fragment>
<Fragment slot="terraform">
<FileTree>
  - packages/common/terraform/src
    - app
      - mcp-servers
        - \<project-name>
          - \<project-name>.tf Module for deploying your MCP Server
    - core
      - agent-core
        - runtime.tf Generic module for deploying to Bedrock AgentCore Runtime
</FileTree>
</Fragment>
</Infrastructure>