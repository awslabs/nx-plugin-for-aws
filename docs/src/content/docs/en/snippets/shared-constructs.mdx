---
title: Shared Constructs
---
import { FileTree } from '@astrojs/starlight/components';
import Infrastructure from '@components/infrastructure.astro';
import Link from '@components/link.astro';

Since this generator vends infrastructure as code based on your chosen `iacProvider`, it will create a project in `packages/common` which includes the relevant CDK constructs or Terraform modules.

The common infrastructure as code project is structured as follows:

<Infrastructure>
<Fragment slot="cdk">
<FileTree>
  - packages/common/constructs
    - src
      - app/ Constructs for infrastructure specific to a project/generator
      - core/ Generic constructs which are reused by constructs in `app`
      - index.ts Entry point exporting constructs from `app`
    - project.json Project build targets and configuration
</FileTree>

:::note
This project is generated using the <Link path="/guides/typescript-project">`ts#project`</Link> generator and therefore configures the same build targets.
:::
</Fragment>
<Fragment slot="terraform">
<FileTree>
  - packages/common/terraform
    - src
      - app/ Terraform modules for infrastructure specific to a project/generator
      - core/ Generic modules which are reused by modules in `app`
    - project.json Project build targets and configuration
</FileTree>

:::note
This project is generated using the <Link path="/guides/terraform-project">`terraform#project`</Link> generator and therefore configures the same build targets.
:::
</Fragment>
</Infrastructure>