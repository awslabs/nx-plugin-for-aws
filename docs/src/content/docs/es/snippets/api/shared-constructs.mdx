---
title: "Construcciones Compartidas de API"
---



import { FileTree } from '@astrojs/starlight/components';
import Infrastructure from '@components/infrastructure.astro';

Para implementar tu API, se generan los siguientes archivos:

<Infrastructure>
<Fragment slot="cdk">
<FileTree>
  - packages/common/constructs/src
    - app
      - apis
        - \<project-name>.ts Constructo de CDK para implementar tu API
    - core
      - api
        - http-api.ts Constructo de CDK para implementar una API HTTP (si seleccionaste implementar una API HTTP)
        - rest-api.ts Constructo de CDK para implementar una API REST (si seleccionaste implementar una API REST)
        - utils.ts Utilidades para los constructos de API
</FileTree>
</Fragment>
<Fragment slot="terraform">
<FileTree>
  - packages/common/terraform/src
    - app
      - apis
        - \<project-name>
          - \<project-name>.tf Módulo para implementar tu API
    - core
      - api
        - http-api
          - http-api.tf Módulo para implementar una API HTTP (si seleccionaste implementar una API HTTP)
        - rest-api
          - rest-api.tf Módulo para implementar una API REST (si seleccionaste implementar una API REST)
</FileTree>
</Fragment>
</Infrastructure>