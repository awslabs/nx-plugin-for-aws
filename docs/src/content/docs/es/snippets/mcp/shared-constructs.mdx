---
title: "Construcciones Compartidas de MCP"
---



import { FileTree } from '@astrojs/starlight/components';
import Infrastructure from '@components/infrastructure.astro';

Para implementar tu MCP Server, se generan los siguientes archivos:

<Infrastructure>
<Fragment slot="cdk">
<FileTree>
  - packages/common/constructs/src
    - app
      - mcp-servers
        - \<project-name>
          - \<project-name>.ts Constructo CDK para implementar tu MCP Server
          - Dockerfile Archivo Docker passthrough utilizado por el constructo CDK
    - core
      - agent-core
        - runtime.ts Constructo CDK genérico para implementar en Bedrock AgentCore Runtime
</FileTree>
</Fragment>
<Fragment slot="terraform">
<FileTree>
  - packages/common/terraform/src
    - app
      - mcp-servers
        - \<project-name>
          - \<project-name>.tf Módulo para implementar tu MCP Server
    - core
      - agent-core
        - runtime.tf Módulo genérico para implementar en Bedrock AgentCore Runtime
</FileTree>
</Fragment>
</Infrastructure>