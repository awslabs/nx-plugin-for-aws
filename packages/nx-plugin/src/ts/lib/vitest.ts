/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { readNxJson, Tree, updateNxJson } from '@nx/devkit';
import { tsquery } from '@phenomnomnominal/tsquery';
import { join } from 'path';
import ts from 'typescript';
import { ConfigureProjectOptions } from './types';
const passWithNoTests = (sourceFile: ts.SourceFile): ts.SourceFile => {
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (
          ts.isCallExpression(node) &&
          ts.isIdentifier(node.expression) &&
          node.expression.text === 'defineConfig' &&
          node.arguments.length === 1
        ) {
          const configObject = node.arguments[0];
          if (ts.isObjectLiteralExpression(configObject)) {
            const testProperty = configObject.properties.find(
              (p) =>
                ts.isPropertyAssignment(p) &&
                ts.isIdentifier(p.name) &&
                p.name.text === 'test'
            );
            if (testProperty && ts.isPropertyAssignment(testProperty)) {
              const testValue = testProperty.initializer;
              if (ts.isObjectLiteralExpression(testValue)) {
                // Check if passWithNoTests already exists
                const hasPassWithNoTests = testValue.properties.some(
                  (p) =>
                    ts.isPropertyAssignment(p) &&
                    ts.isIdentifier(p.name) &&
                    p.name.text === 'passWithNoTests'
                );
                if (!hasPassWithNoTests) {
                  // Add passWithNoTests: true to the test object
                  const newTestValue =
                    context.factory.updateObjectLiteralExpression(testValue, [
                      ...testValue.properties,
                      context.factory.createPropertyAssignment(
                        'passWithNoTests',
                        context.factory.createTrue()
                      ),
                    ]);
                  // Update the test property with the new value
                  const newTestProperty =
                    context.factory.updatePropertyAssignment(
                      testProperty,
                      testProperty.name,
                      newTestValue
                    );
                  // Update the config object with the new test property
                  return context.factory.updateCallExpression(
                    node,
                    node.expression,
                    node.typeArguments,
                    [
                      context.factory.updateObjectLiteralExpression(
                        configObject,
                        configObject.properties.map((p) =>
                          p === testProperty ? newTestProperty : p
                        )
                      ),
                    ]
                  );
                }
              }
            }
          }
        }
        return ts.visitEachChild(node, visit, context);
      }
      return ts.visitNode(rootNode, visit);
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
export const configureVitest = (
  tree: Tree,
  options: ConfigureProjectOptions
) => {
  const configPath = join(options.dir, 'vite.config.ts');
  if (tree.exists(configPath)) {
    const originalSourceFile = tsquery.ast(tree.read(configPath, 'utf-8'));
    let sourceFile = originalSourceFile;
    sourceFile = passWithNoTests(sourceFile);
    const printer = ts.createPrinter({
      removeComments: false,
      newLine: ts.NewLineKind.LineFeed,
    });
    tree.write(
      configPath,
      printer.printNode(ts.EmitHint.Unspecified, sourceFile, originalSourceFile)
    );

    const nxJson = readNxJson(tree);
    updateNxJson(tree, {
      ...nxJson,
      targetDefaults: {
        ...(nxJson.targetDefaults ?? {}),
        test: {
          configurations: {
            'update-snapshot': {
              args: '--update',
            },
          },
        },
      },
    });
  }
};
