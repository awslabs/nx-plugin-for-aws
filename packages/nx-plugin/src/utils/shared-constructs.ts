/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  getPackageManagerCommand,
  joinPathFragments,
  OverwriteStrategy,
  Tree,
} from '@nx/devkit';
import { getNpmScopePrefix, toScopeAlias } from './npm-scope';
import tsProjectGenerator from '../ts/lib/generator';
import { withVersions } from './versions';
import { formatFilesInSubtree } from './format';
import {
  PACKAGES_DIR,
  TYPE_DEFINITIONS_DIR,
  TYPE_DEFINITIONS_NAME,
  SHARED_CONSTRUCTS_DIR,
  SHARED_CONSTRUCTS_NAME,
} from './shared-constructs-constants';

export async function sharedConstructsGenerator(tree: Tree) {
  const npmScopePrefix = getNpmScopePrefix(tree);
  updateGitignore(tree);

  if (
    !tree.exists(
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'project.json'),
    )
  ) {
    await tsProjectGenerator(tree, {
      name: TYPE_DEFINITIONS_NAME,
      directory: PACKAGES_DIR,
      subDirectory: TYPE_DEFINITIONS_DIR,
    });
    tree.delete(joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'src'));
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', TYPE_DEFINITIONS_DIR, 'src'),
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'src'),
      {
        npmScopePrefix,
      },
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', 'common', 'readme'),
      joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR),
      {
        fullyQualifiedName: `${npmScopePrefix}${TYPE_DEFINITIONS_NAME}`,
        name: TYPE_DEFINITIONS_NAME,
        pkgMgrCmd: getPackageManagerCommand().exec,
      },
      {
        overwriteStrategy: OverwriteStrategy.Overwrite,
      },
    );
    await formatFilesInSubtree(tree);
  }
  if (
    !tree.exists(
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    )
  ) {
    await tsProjectGenerator(tree, {
      name: SHARED_CONSTRUCTS_NAME,
      directory: PACKAGES_DIR,
      subDirectory: SHARED_CONSTRUCTS_DIR,
    });
    tree.delete(joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src'));
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR, 'src'),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src'),
      {
        npmScopePrefix,
        scopeAlias: toScopeAlias(npmScopePrefix),
      },
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', 'common', 'readme'),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR),
      {
        fullyQualifiedName: `${npmScopePrefix}${SHARED_CONSTRUCTS_NAME}`,
        name: SHARED_CONSTRUCTS_NAME,
        pkgMgrCmd: getPackageManagerCommand().exec,
      },
      {
        overwriteStrategy: OverwriteStrategy.Overwrite,
      },
    );
    addDependenciesToPackageJson(
      tree,
      withVersions(['constructs', 'aws-cdk-lib']),
      withVersions(['@types/node']),
    );
    await formatFilesInSubtree(tree);
  }
}

const updateGitignore = (tree: Tree) => {
  const gitignore = tree.exists('.gitignore')
    ? tree.read('.gitignore', 'utf-8')
    : '';
  const regex = /runtime-config.json/gm;
  const hasRuntimeConfig = regex.test(gitignore ?? '');
  if (hasRuntimeConfig) {
    return;
  }
  tree.write('.gitignore', `${gitignore}\n\nruntime-config.json`);
};
