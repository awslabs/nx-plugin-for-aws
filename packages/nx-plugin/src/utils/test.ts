/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
