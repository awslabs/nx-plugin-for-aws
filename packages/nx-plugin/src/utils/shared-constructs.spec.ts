/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { describe, expect, it, beforeEach } from 'vitest';
import { Tree, joinPathFragments } from '@nx/devkit';
import { sharedConstructsGenerator } from './shared-constructs';
import {
  PACKAGES_DIR,
  TYPE_DEFINITIONS_DIR,
  SHARED_CONSTRUCTS_DIR,
  TYPE_DEFINITIONS_NAME,
  SHARED_CONSTRUCTS_NAME,
} from './shared-constructs-constants';
import { createTreeUsingTsSolutionSetup, snapshotTreeDir } from './test';

describe('shared-constructs utils', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();

    // Set up npm scope in package.json
    tree.write(
      'package.json',
      JSON.stringify({
        name: '@test-scope/source',
        version: '0.0.0',
        license: 'MIT',
        dependencies: {},
        devDependencies: {},
      }),
    );
  });

  describe('sharedConstructsGenerator', () => {
    it('should generate type definitions when they do not exist', async () => {
      await sharedConstructsGenerator(tree);

      // Check if type definitions project was created
      expect(
        tree.exists(
          joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'project.json'),
        ),
      ).toBeTruthy();

      // Check if type definitions source files were generated
      expect(
        tree.exists(
          joinPathFragments(
            PACKAGES_DIR,
            TYPE_DEFINITIONS_DIR,
            'src',
            'index.ts',
          ),
        ),
      ).toBeTruthy();

      // Take snapshots of the generated source files
      snapshotTreeDir(
        tree,
        joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'src'),
      );
    });

    it('should generate shared constructs when they do not exist', async () => {
      await sharedConstructsGenerator(tree);

      // Check if shared constructs project was created
      expect(
        tree.exists(
          joinPathFragments(
            PACKAGES_DIR,
            SHARED_CONSTRUCTS_DIR,
            'project.json',
          ),
        ),
      ).toBeTruthy();

      // Check if shared constructs source files were generated
      expect(
        tree.exists(
          joinPathFragments(
            PACKAGES_DIR,
            SHARED_CONSTRUCTS_DIR,
            'src',
            'index.ts',
          ),
        ),
      ).toBeTruthy();

      // Check if core files were generated
      expect(
        tree.exists(
          joinPathFragments(
            PACKAGES_DIR,
            SHARED_CONSTRUCTS_DIR,
            'src',
            'core',
            'app.ts',
          ),
        ),
      ).toBeTruthy();

      // Take snapshots of the generated source files
      snapshotTreeDir(
        tree,
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src'),
      );
    });

    it('should add required dependencies when generating shared constructs', async () => {
      await sharedConstructsGenerator(tree);

      // Read package.json and check if dependencies were added
      const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));

      expect(packageJson.dependencies).toHaveProperty('constructs');
      expect(packageJson.dependencies).toHaveProperty('aws-cdk-lib');
      expect(packageJson.devDependencies).toHaveProperty('@types/node');
    });

    it('should not generate type definitions when they already exist', async () => {
      // Create type definitions project first
      tree.write(
        joinPathFragments(PACKAGES_DIR, TYPE_DEFINITIONS_DIR, 'project.json'),
        JSON.stringify({
          name: TYPE_DEFINITIONS_NAME,
          sourceRoot: joinPathFragments(
            PACKAGES_DIR,
            TYPE_DEFINITIONS_DIR,
            'src',
          ),
        }),
      );

      // Create a marker file to check if it gets overwritten
      const markerFilePath = joinPathFragments(
        PACKAGES_DIR,
        TYPE_DEFINITIONS_DIR,
        'marker.txt',
      );
      tree.write(markerFilePath, 'This is a marker file');

      await sharedConstructsGenerator(tree);

      // Check if marker file still exists (meaning the directory wasn't recreated)
      expect(tree.exists(markerFilePath)).toBeTruthy();
    });

    it('should not generate shared constructs when they already exist', async () => {
      // Create shared constructs project first
      tree.write(
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
        JSON.stringify({
          name: SHARED_CONSTRUCTS_NAME,
          sourceRoot: joinPathFragments(
            PACKAGES_DIR,
            SHARED_CONSTRUCTS_DIR,
            'src',
          ),
        }),
      );

      // Create a marker file to check if it gets overwritten
      const markerFilePath = joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'marker.txt',
      );
      tree.write(markerFilePath, 'This is a marker file');

      await sharedConstructsGenerator(tree);

      // Check if marker file still exists (meaning the directory wasn't recreated)
      expect(tree.exists(markerFilePath)).toBeTruthy();
    });
  });
});
