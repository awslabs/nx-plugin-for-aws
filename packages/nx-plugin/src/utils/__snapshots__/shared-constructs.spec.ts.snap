// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shared-constructs utils > sharedConstructsGenerator > should generate shared constructs when they do not exist > packages/common/constructs/src/app/index.ts 1`] = `""`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate shared constructs when they do not exist > packages/common/constructs/src/core/app.ts 1`] = `
"import { App as _App, AppProps, Aspects, IAspect, Stack } from 'aws-cdk-lib';
import { IConstruct } from 'constructs';

export class App extends _App {
  constructor(props?: AppProps) {
    super(props);

    Aspects.of(this).add(new MetricsAspect());
  }
}

/**
 * Adds information to CloudFormation stack descriptions to provide usage metrics for @aws/nx-plugin
 */
class MetricsAspect implements IAspect {
  visit(node: IConstruct): void {
    if (node instanceof Stack) {
      const id = '';
      const version = '';
      const tags: string[] = [];
      node.templateOptions.description =
        \`\${node.templateOptions.description ?? ''} (\${id}) (version:\${version}) (tags:\${tags.join(',')})\`.trim();
    }
  }
}
"
`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate shared constructs when they do not exist > packages/common/constructs/src/core/index.ts 1`] = `
"export * from './app.js';
export * from './runtime-config.js';
"
`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate shared constructs when they do not exist > packages/common/constructs/src/core/runtime-config.ts 1`] = `
"import type { IRuntimeConfig } from ':test-scope/common-types';
import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

const RuntimeConfigKey = '__RuntimeConfig__';

export class RuntimeConfig extends Construct {
  private readonly _runtimeConfig: Partial<IRuntimeConfig> = {};

  static ensure(scope: Construct): RuntimeConfig {
    const stack = Stack.of(scope);
    return (
      RuntimeConfig.of(scope) ?? new RuntimeConfig(stack, RuntimeConfigKey)
    );
  }

  static of(scope: Construct): RuntimeConfig | undefined {
    const stack = Stack.of(scope);
    return stack.node.tryFindChild(RuntimeConfigKey) as
      | RuntimeConfig
      | undefined;
  }

  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  get config(): Partial<IRuntimeConfig> {
    return this._runtimeConfig;
  }
}
"
`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate shared constructs when they do not exist > packages/common/constructs/src/index.ts 1`] = `
"export * from './app/index.js';
export * from './core/index.js';
"
`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate type definitions when they do not exist > packages/common/types/src/index.ts 1`] = `
"export * from './runtime-config.js';
"
`;

exports[`shared-constructs utils > sharedConstructsGenerator > should generate type definitions when they do not exist > packages/common/types/src/runtime-config.ts 1`] = `
"// eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-empty-interface
export interface IRuntimeConfig {}
"
`;
